toolchain/check/impl_lookup.cpp

```diff
@@ -865,9 +892,30 @@ static auto CollectCandidateImplsForQuery(
   // TODO: Allow Carbon code to provide a priority ordering explicitly. For
   // now they have all the same priority, so the priority is the order in
   // which they are found in code.
-  llvm::stable_sort(candidate_impls, compare);
+  llvm::stable_sort(candidates.impls, compare);
 
-  return candidate_impls;
+  return candidates;
+}
+
+// Given a value that is either a type or a non-type facet, returns the
+// corresponding type.
+static auto GetFacetAsType(Context& context, SemIR::LocId loc_id,
+                           SemIR::ConstantId facet_or_type_const_id)
+    -> SemIR::TypeId {
+  auto facet_or_type_id =
+      context.constant_values().GetInstId(facet_or_type_const_id);
+  auto type_type_id = context.insts().Get(facet_or_type_id).type_id();
+  CARBON_CHECK(context.types().IsFacetType(type_type_id) ||
+               type_type_id == SemIR::ErrorInst::TypeId);
+
+  if (context.types().Is<SemIR::FacetType>(type_type_id)) {
+    // It's a facet; access its type.
+    facet_or_type_id = GetOrAddInst<SemIR::FacetAccessType>(
+        context, loc_id,
+        {.type_id = SemIR::TypeType::TypeId,
+         .facet_value_inst_id = facet_or_type_id});
+  }
+  return context.types().GetTypeIdForTypeInstId(facet_or_type_id);
 }
```

Comment:
I guess this is a cheaper version of calling `ExprAsType`, with CHECKs involved, right? I had thought of replacing some other places that construct `FacetAccessType` with `ExprAsType`. Like

https://github.com/carbon-language/carbon-lang/blob/a179bd461b4a07bf2defe91e68f1bb9edfa60551/toolchain/check/handle_require.cpp#L74C1-L77C1

Maybe this should go live beside `ExprAsType` so we can reuse it?